src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/TextDocumentUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,225 @@
+            if (componentTagText is null)
+            {
+                return null;
+            }
+
+            return new WorkspaceEdit
+            {
+                DocumentChanges = new Container<WorkspaceEditDocumentChange>(
+                    new WorkspaceEditDocumentChange(
+                        new TextDocumentEdit
+                        {
+                            TextDocument = new()
+                            {
+                                Uri = request.TextDocument.Uri
+                            },
+                            Edits = new[]
+                            {
+                                    new TextEdit
+                                    {
+                                        NewText = componentTagText,
+                                        Range = request.Range
+                                    }
+                            }
+                        }
+                    )
+                )
+            };
+        }
+
+        private async Task<string?> TryGetComponentTagAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Trying to find document info for dropped uri {uri}.");
+
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var descriptor = await _razorComponentSearchEngine.TryGetTagHelperDescriptorAsync(documentSnapshot).ConfigureAwait(false);
+            if (descriptor is null)
+            {
+                _logger.LogWarning($"Failed to find tag helper descriptor.");
+                return null;
+            }
+
+            var typeName = descriptor.GetTypeNameIdentifier();
+            if (string.IsNullOrWhiteSpace(typeName))
+            {
+                _logger.LogWarning($"Found a tag helper, {descriptor.Name}, but it has an empty TypeNameIdentifier.");
+                return null;
+            }
+
+            // TODO: Add @using statements if required, or fully qualify (GetTypeName())
+
+            var sb = new StringBuilder();
+            sb.Append('<');
+            sb.Append(typeName);
+
+            foreach (var requiredAttribute in descriptor.EditorRequiredAttributes)
+            {
+                sb.Append(' ');
+                sb.Append(requiredAttribute.Name);
+                sb.Append("=\"\"");
+            }
+
+            if (descriptor.AllowedChildTags.Count > 0)
```

Comment:
Is AllowedChildTags the thing to check here? My understanding is that if that's not supplied then ANY tag is legal rather than NO tag. I thought there was something like `descriptor.IsSelfClosing` which might be more appropriate?

Reply:
I have no idea! There is no `IsSelfClosing` or similar that I could find. Could always just insert the long form I guess.